src/human_encoding/parse/mod.rs

```diff
@@ -0,0 +1,629 @@
+            let right = data
+                .right_index
+                .and_then(|idx| converted[idx].as_ref().map(Arc::clone));
+
+            let maybe_inner = match data.node.inner {
+                ResolvedInner::Missing { ref name, .. } => {
+                    errors.add(data.node.position, Error::NameMissing(Arc::clone(name)));
+                    None
+                }
+                ResolvedInner::NoExpr { ref name, .. } => {
+                    errors.add(data.node.position, Error::NameIncomplete(Arc::clone(name)));
+                    None
+                }
+                ResolvedInner::Reference(..) => {
+                    converted.push(left);
+                    continue;
+                }
+                ResolvedInner::AssertL(..) => left.zip(right).map(|(left, right)| {
+                    let cmr = right.cmr();
+                    node::Inner::AssertL(left, cmr)
+                }),
+                ResolvedInner::AssertR(..) => left.zip(right).map(|(left, right)| {
+                    let cmr = left.cmr();
+                    node::Inner::AssertR(cmr, right)
+                }),
+                ResolvedInner::Inline(ref inner) => inner
+                    .as_ref()
+                    .map_left_right(|_| left, |_| right)
+                    .copy_disconnect()
+                    .copy_witness()
+                    .transpose(),
+            };
+
+            let inner = match maybe_inner {
+                Some(inner) => inner,
+                None => {
+                    converted.push(None);
+                    continue;
+                }
+            };
+
+            let name = data
+                .node
+                .name
+                .as_ref()
+                .map(Arc::clone)
+                .unwrap_or_else(|| Arc::from(namer.assign_name(inner.as_ref()).as_str()));
+
+            let node = NamedConstructNode::new(
+                Arc::clone(&name),
+                data.node.position,
+                Arc::clone(&data.node.user_source_types),
+                Arc::clone(&data.node.user_target_types),
+                inner,
+            );
+
+            match node {
+                Ok(node) => {
+                    let node = Arc::new(node);
+                    converted.push(Some(node));
+                }
+                Err(e) => {
+                    errors.add(data.node.position, e);
+                    converted.push(None);
+                }
+            }
+        }
+
+        if let Some(Some(root)) = converted.pop() {
+            let finalized = if &**name == "main" {
+                root.finalize_types_main()
+            } else {
+                root.finalize_types_non_main()
+            };
+
+            match finalized {
+                Ok(root) => {
+                    roots.insert(Arc::clone(name), root);
+                }
+                Err(errs) => errors.merge(&errs),
+            }
+        }
+    }
+
+    // Finally, build name map.
```

Comment:
8e711fd6b8a34c454d7ec83a8ad672cc6a8e12f1: Outdated comment. The map is already built.